pimbd/files/pimbd.init

```diff
@@ -4,18 +4,148 @@ START=90
 STOP=10
 USE_PROCD=1
 
+TYPE=''
+
+DEV=''
+PIM='off'
+SSBIDIR='off'
+MLD='off'
+IGMP='off'
+
+RPAADDRESS=''
+RPALIST=''
+
+reset_link_values() {
+    PIM='off'
+    SSBIDIR='off'
+    MLD='off'
+    IGMP='off'
+}
+
+apply_link() {
+    if test ! -z "$DEV"; then
+        echo "Setting pimbc link set $DEV pim $PIM ssbidir $SSBIDIR mld $MLD igmp $IGMP"
+        echo "link set $DEV pim $PIM ssbidir $SSBIDIR mld $MLD igmp $IGMP" >> /tmp/pimbd.conf
+    fi
+    reset_link_values
+}
+
+apply_rpa() {
+    if test -z "$RPAADDRESS"; then
+        echo "No RPA address is set!"
+        return
+    fi
+    for i in $RPALIST; do
+        echo "adding pimbc rpa add $RPAADDRESS $i"
+        echo "rpa add $RPAADDRESS $i" >> /tmp/pimbd.conf
+    done
+    RPAADDRESS=''
+    RPALIST=''
+}
+
+store_interface_option() {
+    local name="$1"
+    local value="$2"
+    case $name in
+    dev)
+        DEV=$value
+        ;;
+    pim)
+        PIM=$value
+        ;;
+    ssbidir)
+        SSBIDIR=$value
+        ;;
+    mld)
+        MLD=$value
+        ;;
+    igmp)
+        IGMP=$value
+        ;;
+    esac
+}
+
+store_rpa_option() {
+    local name="$1"
+    local value="$2"
+    echo "store_rpa_option $name $value"
+    case $name in
+    address)
+        RPAADDRESS="$value"
+        ;;
+    esac
+}
+
+store_rpa_list() {
+    local name="$1"
+    local value="$2"
+    echo "store rpa_list $name $value"
+    case $name in
+    prefix)
+        RPALIST="${RPALIST} $value"
+        ;;
+    esac
+}
+
 start_service() {
     . /lib/functions.sh
     . /lib/functions/network.sh
+
+    rm -f /tmp/pimbd.conf
+
+    config_cb() {
+        local name="$1"
+        local value="$2"
+        # commands to be run for every option
+        echo "config_cb $name $value"
+        case $TYPE in
+            interface)
+            apply_link
+            ;;
+        rpa)
+            apply_rpa
+            ;;
+        esac
+        TYPE=$name
+        DEV=$value
+    }
+
+    option_cb() {
+        local name="$1"
+        local value="$2"
+        # commands to be run for every option
+        echo "option_cb $name $value"
+        case $TYPE in
+            interface)
+                store_interface_option "$name" "$value"
+                ;;
+            rpa)
+                store_rpa_option "$name" "$value"
+                ;;
+        esac
+    }
+
+    list_cb() {
+        local name="$1"
+        local value="$2"
+        # commands to be run for every list item
+        echo "list_cb $name $value"
+
+        case $TYPE in
+            rpa)
+                store_rpa_list "$name" "$value"
+                ;;
+        esac
+    }
     config_load pimb
 
     procd_open_instance
     procd_set_param command /usr/sbin/pimbd
     procd_append_param command -S
     procd_append_param command -L 6
+    procd_append_param command -c /tmp/pimbd.conf
```

Comment:
Shouldn't this be used only if the file is actually written?  I'm assuming that having a file name here means that the default config file will not be loaded.  If this is indeed true, it would break existing config.

Reply:
Without this patch the pimbd does not load any configuration file at all. The pimbd just starts and doesn't do anything. It must be configured by using one of the client tools like pimbc. This has to be done after every reboot of the device as well as after every restart of the daemon. There seems to be no configuration at all. I haven't found one in the pimbd.c file.

The file /tmp/pimbd.conf gets only written if there is a configuration in /etc/config/pimb. If the file dosn't exist the pimbd starts and runs as it would do without the patch. This should ensure backward compatibility if an user has some custom code in place.

If there is a configuration in /etc/config/pimb the file /tmp/pimbd.conf will be written and the pimbd will load the configuration.